test/e2e/httpproxy/fqdn_test.go

```diff
@@ -81,4 +83,99 @@ func testWildcardFQDN(namespace string) {
 		err := f.CreateHTTPProxy(p)
 		require.NotNil(t, err, "Expected invalid wildcard to be rejected.")
 	})
+
+	Specify("wildcard routing works", func() {
+		t := f.T()
+
+		f.Fixtures.Echo.Deploy(namespace, "echo-slash-default")
+		f.Fixtures.Echo.Deploy(namespace, "ingress-conformance-echo")
+		f.Fixtures.Echo.Deploy(namespace, "ingress-conformance-bar")
+
+		proxyWildcard := &contourv1.HTTPProxy{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "wildcard",
+			},
+			Spec: contourv1.HTTPProxySpec{
+				VirtualHost: &contourv1.VirtualHost{
+					Fqdn: "*.projectcontour.io",
+				},
+				Routes: []contourv1.Route{{
+					Services: []contourv1.Service{
+						{
+							Name: "echo-slash-default",
+							Port: 80,
+						},
+					},
+				}},
+			},
+		}
+		proxyFullFQDN := &contourv1.HTTPProxy{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "full-fqdn",
+			},
+			Spec: contourv1.HTTPProxySpec{
+				VirtualHost: &contourv1.VirtualHost{
+					Fqdn: "projectcontour.io",
+				},
+				Routes: []contourv1.Route{{
+					Services: []contourv1.Service{
+						{
+							Name: "ingress-conformance-echo",
+							Port: 80,
+						},
+					},
+				}},
+			},
+		}
+		proxyFullFQDNSubdomain := &contourv1.HTTPProxy{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "fqdn-subdomain",
+			},
+			Spec: contourv1.HTTPProxySpec{
+				VirtualHost: &contourv1.VirtualHost{
+					Fqdn: "bar.projectcontour.io",
+				},
+				Routes: []contourv1.Route{{
+					Services: []contourv1.Service{
+						{
+							Name: "ingress-conformance-bar",
+							Port: 80,
+						},
+					},
+				}},
+			},
+		}
+		f.CreateHTTPProxyAndWaitFor(proxyWildcard, httpProxyValid)
+		f.CreateHTTPProxyAndWaitFor(proxyFullFQDN, httpProxyValid)
+		f.CreateHTTPProxyAndWaitFor(proxyFullFQDNSubdomain, httpProxyValid)
+
+		cases := map[string]string{
+			"projectcontour.io":                         "ingress-conformance-echo",
+			"www.projectcontour.io":                     "echo-slash-default",
+			"bar.projectcontour.io":                     "ingress-conformance-bar",
+			"foo.projectcontour.io":                     "echo-slash-default",
+			"foo.bar.projectcontour.io":                 "echo-slash-default",
+			"i.n.g.r.e.s.s.r.u.l.e.s.projectcontour.io": "echo-slash-default",
```

Comment:
do we want this to match the wildcard? that would be different behavior than Ingress and Gateway API which only allow a single DNS label to match: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Hostname

we used this to in the Ingress and Gateway API processors: https://github.com/projectcontour/contour/blob/0d824ba6d0bf3a4a97ee2024b4a45fea21b278ec/internal/dag/ingress_processor.go#L263-L278

Reply:
yeah can match that, no reason to deviate.

Reply:
you're ruining my fun @sunjayBhatia! :grin: